src/common.jl

```diff
@@ -48,3 +48,112 @@ parameterless_type(x) = parameterless_type(typeof(x))
 parameterless_type(x::Type) = __parameterless_type(x)
 
 @non_differentiable adapt_randn(::Any...)
+
+"""
+    make_closure(f, g)
+
+Return a closure of the form `(x, args...) -> f(g(args...), x)`.
+
+# Examples
+
+This is particularly useful when one wants to avoid broadcasting over constructors
+which can sometimes cause issues with type-inference, in particular when combined
+with reverse-mode AD frameworks.
+
+```juliarepl
+julia> using DistributionsAD, Distributions, ReverseDiff, BenchmarkTools
+
+julia> const data = randn(1000);
+
+julia> x = randn(length(data));
+
+julia> f(x) = sum(logpdf.(Normal.(x), data))
+f (generic function with 2 methods)
+
+julia> @btime ReverseDiff.gradient(\$f, \$x);
+  848.759 μs (14605 allocations: 521.84 KiB)
+
+julia> # Much faster with ReverseDiff.jl.
+       g(x) = let g_inner = DistributionsAD.make_closure(logpdf, Normal)
+           sum(g_inner.(data, x))
+       end
+g (generic function with 1 method)
+
+julia> @btime ReverseDiff.gradient(\$g, \$x);
+  17.460 μs (17 allocations: 71.52 KiB)
+```
+
+See https://github.com/TuringLang/Turing.jl/issues/1934 more further discussion.
+
+# Notes
+To really go "vrooom!\" one needs to specialize on the arguments, e.g. if one
+has a function `myfunc` then we need to define
+
+```julia
+make_closure(::typeof(myfunc), ::Type{D}) where {D} = myfunc(D(args...), x)
+```
+
+This can also be done using `DistributionsAD.@specialize_make_closure`:
+
+```julia
+julia> mylogpdf(d, x) = logpdf(d, x)
+mylogpdf (generic function with 1 method)
+
+julia> h(x) = let inner = DistributionsAD.make_closure(mylogpdf, Normal)
+           sum(inner.(data, x))
+       end
+h (generic function with 1 method)
+
+julia> @btime ReverseDiff.gradient(\$h, \$x);
+  1.220 ms (37011 allocations: 1.42 MiB)
+
+julia> DistributionsAD.@specialize_make_closure mylogpdf
+
+julia> @btime ReverseDiff.gradient(\$h, \$x);
+  17.038 μs (17 allocations: 71.52 KiB)
+```
+"""
+make_closure(f, g) = (x, args...) -> f(g(args...), x)
+make_closure(f, ::Type{D}) where {D} = (x, args...) -> f(D(args...), x)
```

Comment:
I wonder if there are any possible performance/compiler benefits by not closing over the variables but to use (more-Julian) callable structs that capture `f` and `g`. In any case, I think you want
```suggestion
make_closure(f::F, g::G) where {F,G} = (x, args...) -> f(g(args...), x)
make_closure(f::F, ::Type{D}) where {F,D} = (x, args...) -> f(D(args...), x)
```

Reply:
I unfortuantely tried that but AFAIK it ends up with this issue of closing over a `UnionAll` type again, which is exactly what we're trying to avoid (because of the issues it's causing with some AD backends) :confused: 

I might have not done it correctly though.

But you suggestion I _have_ tried, and it unfortunately doesn't have an affect. If you just look at the returned closures, they're all the same one :confused:

Reply:
Maybe something like
```julia
struct Closure{F,G} end

Closure(::F, ::G) where {F,G} = Closure{F,G}()
Closure(::F, ::Type{G}) where {F,G} = Closure{F,G}()
Closure(::Type{F}, ::G) where {F,G} = Closure{F,G}()
Closure(::Type{F}, ::Type{G}) where {F,G} = Closure{F,G}()

for f in [pdf, logpdf, cdf, logcdf]
    @eval (::$(Closure){typeof($f),G})(x, args...) where {G} = $f(G(args...), x)
end
```
?

Reply:
```julia
Closure(::F, ::Type{G}) where {F,G} = Closure{F,G}()
```
and others avoid the `UnionAll` issue.

Reply:
Have a look at what I've done now :+1: